libs/zzz/solver/src/calc.ts

```diff
@@ -13,22 +17,58 @@ import {
 import type { Stats } from '@genshin-optimizer/zzz/db'
 import type { DiscStats } from './common'
 
+export function passSetFilter(
+  discs: DiscStats[],
+  filter2: DiscSetKey[],
+  filter4: DiscSetKey[]
+): boolean {
+  const setCount: Partial<Record<DiscSetKey, number>> = {}
+  if (!filter4.length && !filter2.length) return true
+  for (const { setKey } of discs) setCount[setKey] = (setCount[setKey] || 0) + 1
+
+  if (filter4.length) {
+    let has4 = false
+    for (const key in setCount) {
```

Comment:
can do `Object.entries(setCount).forEach(([key, count]) => {})` and it will have proper typings and no need for `!` access

Reply:
Yes, but its less performant, since it creates a new array